tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/AbstractChouOrlandiOT.java

```diff
@@ -0,0 +1,187 @@
+
+    /**
+     * multiplies the BigInteger with the Generator
+     * @param input
+     * @return a new Element T
+     */
+    abstract T multiplyWithGenerator(BigInteger input);
+
+    public AbstractChouOrlandiOT(int otherId, Drbg randBit, Network network) {
+        this.otherId = otherId;
+        this.network = network;
+        this.randNum = new DrngImpl(randBit);
+        this.hashDigest = ExceptionConverter.safe(() -> MessageDigest.getInstance(HASH_ALGORITHM),
+                "Missing secure, hash function which is dependent in this library");
+    }
+
+    @Override
+    public void send(StrictBitVector messageZero, StrictBitVector messageOne) {
+        int maxBitLength = Math.max(messageZero.getSize(), messageOne.getSize());
+        Pair<byte[], byte[]> seedMessages = sendRandomOt();
+        byte[] encryptedZeroMessage = PseudoOtp.encrypt(messageZero.toByteArray(),
+                seedMessages.getFirst(), maxBitLength / Byte.SIZE);
+        byte[] encryptedOneMessage = PseudoOtp.encrypt(messageOne.toByteArray(),
+                seedMessages.getSecond(), maxBitLength / Byte.SIZE);
+        network.send(otherId, encryptedZeroMessage);
+        network.send(otherId, encryptedOneMessage);
+    }
+
+    @Override
+    public StrictBitVector receive(boolean choiceBit) {
+        byte[] seed = receiveRandomOt(choiceBit);
+        byte[] encryptedZeroMessage = network.receive(otherId);
+        byte[] encryptedOneMessage = network.receive(otherId);
+
+        if (encryptedZeroMessage.length != encryptedOneMessage.length) {
+            throw new MaliciousException("The length of the two choice messages is not equal");
+        }
+        byte[] unpaddedMessage;
+        if (choiceBit == false) {
+            unpaddedMessage = PseudoOtp.decrypt(encryptedZeroMessage, seed);
+        } else {
+            unpaddedMessage = PseudoOtp.decrypt(encryptedOneMessage, seed);
+        }
+        return new StrictBitVector(unpaddedMessage);
+    }
+
+    /**
+     * Completes the receiver's part of the Chou-Orlandi OT in order to receive a random message of the
+     * length of hash digest.
+     *
+     * @return The random message received
+     */
+    private byte[] receiveRandomOt(boolean choiceBit) {
+        // S
+        T S = this.decodeElement(network.receive(otherId));
+        // T = G(S)
+        T T = hashToFieldElement(S, "FRESCO|ChouOrlandi|OT");
+        // R = T^c*g^x
+        // if c = 0 -> R = g^x
+        // if c = 1 -> R = T*g^x
+        BigInteger x = randNum.nextBigInteger(getDhModulus());
+        T R;
+        if (choiceBit == false) {
+            R = T.exponentiation(BigInteger.ZERO).groupOp(getGenerator().exponentiation(x));
+        } else {
+            R = T.exponentiation(BigInteger.ONE).groupOp(getGenerator().exponentiation(x));
+        }
+        network.send(otherId, R.toByteArray());
+
+        // k = H(S, R, S^x)
+        byte[] key;
+        hashDigest.update(S.toByteArray());
```

Comment:
I'm not sure I understand by hashing locally?

Reply:
Sorry, badly written from my side. The we need to ensure that what gets returned by `toByteArray` is always the same amount of bytes, regardless of the magnitude of the number it represents. Since it is most likely big endian this is a problem, and thus the returned value need to be prepended by empty bytes if S is not a number that is large enough to use the maximum amount of bytes

Reply:
Solved by 0375fd3